'Expected expression' error in opencl tests due to undefined TEST_DIR

When building with `BUILD_TEST` option `ON` in Xcode, I get multiple `Expected expression` errors in lines which use the `TEST_DIR` macro. Detailed error message for `moments_opencl`, and similar errors for `harris_opencl`:

```
CompileC Xcode/test/ARRAYFIRE.build/Release/moments_opencl.build/Objects-normal/x86_64/moments.o test/moments.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler
    cd /Users/as12312/Software/ArrayFire
    export LANG=en_US.US-ASCII
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -x c++ -arch x86_64 -fmessage-length=0 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit=0 -Wno-trigraphs -fpascal-strings -O3 -Wno-missing-field-initializers -Wno-missing-prototypes -Wno-return-type -Wno-non-virtual-dtor -Wno-overloaded-virtual -Wno-exit-time-destructors -Wno-missing-braces -Wparentheses -Wswitch -Wno-unused-function -Wno-unused-label -Wno-unused-parameter -Wno-unused-variable -Wunused-value -Wno-empty-body -Wno-uninitialized -Wno-unknown-pragmas -Wno-shadow -Wno-four-char-constants -Wno-conversion -Wno-constant-conversion -Wno-int-conversion -Wno-bool-conversion -Wno-enum-conversion -Wno-shorten-64-to-32 -Wno-newline-eof -Wno-c++11-extensions -DCMAKE_INTDIR=\"Release\" -DWITH_FREEIMAGE -DGLEW_MX -DWITH_GRAPHICS -DOS_MAC -DAF_UNIFIED -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -fasm-blocks -fstrict-aliasing -Wdeprecated-declarations -Winvalid-offsetof -mmacosx-version-min=10.11 -Wno-sign-conversion -I/Users/as12312/Software/ArrayFire/Xcode/test/Release/include -I/usr/local/include -I/Users/as12312/Software/ArrayFire/Xcode/third_party/forge/include -I/usr/X11R6/include -I/Users/as12312/Software/ArrayFire/include -I/Users/as12312/Software/ArrayFire/src/backend -I/Users/as12312/Software/ArrayFire/src/api/c -I/Users/as12312/Software/ArrayFire/test/../test/gtest/include -I/Users/as12312/Software/ArrayFire/test -I/Users/as12312/Software/ArrayFire/Xcode/test/ARRAYFIRE.build/Release/moments_opencl.build/DerivedSources/x86_64 -I/Users/as12312/Software/ArrayFire/Xcode/test/ARRAYFIRE.build/Release/moments_opencl.build/DerivedSources -Wmost -Wno-four-char-constants -Wno-unknown-pragmas -F/Users/as12312/Software/ArrayFire/Xcode/test/Release -F/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks -DNDEBUG -DAF_OPENCL -Wall -fvisibility=hidden -D TEST_DIR=\\/Users/as12312/Software/ArrayFire/test/data\\ -MMD -MT dependencies -MF /Users/as12312/Software/ArrayFire/Xcode/test/ARRAYFIRE.build/Release/moments_opencl.build/Objects-normal/x86_64/moments.d --serialize-diagnostics /Users/as12312/Software/ArrayFire/Xcode/test/ARRAYFIRE.build/Release/moments_opencl.build/Objects-normal/x86_64/moments.dia -c /Users/as12312/Software/ArrayFire/test/moments.cpp -o /Users/as12312/Software/ArrayFire/Xcode/test/ARRAYFIRE.build/Release/moments_opencl.build/Objects-normal/x86_64/moments.o

/Users/as12312/Software/ArrayFire/test/moments.cpp:145:31: error: expected expression
    momentsOnImageTest(string(TEST_DIR"/moments/gray_seq_16_moments.test"), string(TEST_DIR"/imageio/gray_seq_16.png"), false);
                              ^
In file included from <built-in>:337:
<command line>:9:18: note: expanded from here
#define TEST_DIR \/Users/as12312/Software/ArrayFire/test/data\\
                 ^
/Users/as12312/Software/ArrayFire/test/moments.cpp:145:84: error: expected expression
    momentsOnImageTest(string(TEST_DIR"/moments/gray_seq_16_moments.test"), string(TEST_DIR"/imageio/gray_seq_16.png"), false);
                                                                                   ^
In file included from <built-in>:337:
<command line>:9:18: note: expanded from here
#define TEST_DIR \/Users/as12312/Software/ArrayFire/test/data\\
                 ^
/Users/as12312/Software/ArrayFire/test/moments.cpp:150:31: error: expected expression
    momentsTest<float>(string(TEST_DIR"/moments/simple_mat_batch_moments.test"));
                              ^
In file included from <built-in>:337:
<command line>:9:18: note: expanded from here
#define TEST_DIR \/Users/as12312/Software/ArrayFire/test/data\\
                 ^
/Users/as12312/Software/ArrayFire/test/moments.cpp:155:31: error: expected expression
    momentsOnImageTest(string(TEST_DIR"/moments/color_seq_16_moments.test"), string(TEST_DIR"/imageio/color_seq_16.png"), true);
                              ^
In file included from <built-in>:337:
<command line>:9:18: note: expanded from here
#define TEST_DIR \/Users/as12312/Software/ArrayFire/test/data\\
                 ^
/Users/as12312/Software/ArrayFire/test/moments.cpp:155:85: error: expected expression
    momentsOnImageTest(string(TEST_DIR"/moments/color_seq_16_moments.test"), string(TEST_DIR"/imageio/color_seq_16.png"), true);
                                                                                    ^
In file included from <built-in>:337:
<command line>:9:18: note: expanded from here
#define TEST_DIR \/Users/as12312/Software/ArrayFire/test/data\\
                 ^
/Users/as12312/Software/ArrayFire/test/moments.cpp:160:35: error: expected expression
    momentsTest<TypeParam>(string(TEST_DIR"/moments/simple_mat_moments.test"));
                                  ^
In file included from <built-in>:337:
<command line>:9:18: note: expanded from here
#define TEST_DIR \/Users/as12312/Software/ArrayFire/test/data\\
                 ^
6 errors generated.
```

--- a/CMakeModules/build_forge.cmake
+++ b/CMakeModules/build_forge.cmake
@@ -2,13 +2,31 @@ INCLUDE(ExternalProject)
 
 SET(prefix ${CMAKE_BINARY_DIR}/third_party/forge)
 
+IF(CMAKE_GENERATOR MATCHES "Xcode") # TODO: Also for "Visual Studio"?
+    # FIXME: Cannot use $<CONFIG> generator expression here because add_custom_command
+    #        does not yet support it for the OUTPUT argument, see also:
+    #        - Old "duplicate":   https://cmake.org/Bug/view.php?id=12877
+    #        - Old issue tracker: https://cmake.org/Bug/view.php?id=13840
+    #        - New issue tracker: https://gitlab.kitware.com/cmake/cmake/issues/13840
+    #        In the meantime, use CMAKE_BUILD_TYPE if set by user, assuming that it
+    #        is the primary build configuration used. Otherwise, default to Release.
+    IF(CMAKE_BUILD_TYPE)
+        SET(forge_lib_config ${CMAKE_BUILD_TYPE})
+    ELSE()
+        SET(forge_lib_config Release)
+    ENDIF()
+    SET(forge_lib_infix "/${forge_lib_config}")
+ELSE()
+    SET(forge_lib_config "$<CONFIG>")
+    SET(forge_lib_infix)
+ENDIF()
 IF(WIN32)
     SET(forge_lib_prefix "${prefix}/lib")
 ELSE(WIN32)
     SET(forge_lib_prefix "${prefix}/src/forge-ext-build/src")
 ENDIF(WIN32)
 
-SET(forge_location "${forge_lib_prefix}/${CMAKE_SHARED_LIBRARY_PREFIX}forge${CMAKE_SHARED_LIBRARY_SUFFIX}")
+SET(forge_location "${forge_lib_prefix}${forge_lib_infix}/${CMAKE_SHARED_LIBRARY_PREFIX}forge${CMAKE_SHARED_LIBRARY_SUFFIX}")
 IF(CMAKE_VERSION VERSION_LESS 3.2)
     IF(CMAKE_GENERATOR MATCHES "Ninja")
         MESSAGE(WARNING "Building forge with Ninja has known issues with CMake older than 3.2")
@@ -36,6 +54,7 @@ ExternalProject_Add(
     -DGLFW_ROOT_DIR:STRING=${GLFW_ROOT_DIR}
     -DCMAKE_INSTALL_PREFIX:PATH=<INSTALL_DIR>
     -DBUILD_EXAMPLES:BOOL=OFF
+    BUILD_COMMAND ${CMAKE_COMMAND} --build . --config ${forge_lib_config}
     ${byproducts}
     )
 
@@ -46,7 +65,7 @@ SET_TARGET_PROPERTIES(forge PROPERTIES IMPORTED_LOCATION ${forge_location})
 IF(WIN32)
     SET_TARGET_PROPERTIES(forge PROPERTIES IMPORTED_IMPLIB ${forge_lib_prefix}/forge.lib)
 ELSE(WIN32)
-    SET(forge_bindir_location ${binary_dir}/src/${CMAKE_SHARED_LIBRARY_PREFIX}forge${CMAKE_SHARED_LIBRARY_SUFFIX})
+    SET(forge_bindir_location ${binary_dir}/src${forge_lib_infix}/${CMAKE_SHARED_LIBRARY_PREFIX}forge${CMAKE_SHARED_LIBRARY_SUFFIX})
     IF(NOT (${forge_bindir_location} STREQUAL ${forge_location}))
         MESSAGE(WARNING "Did the forge binary location move? (Have ${forge_bindir_location} vs ${forge_location})")
     ENDIF()

--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -139,11 +139,13 @@ ELSE(${USE_RELATIVE_TEST_DIR})  # Not using relative test data directory
     SET(TESTDATA_SOURCE_DIR "${CMAKE_CURRENT_SOURCE_DIR}/data")
 ENDIF(${USE_RELATIVE_TEST_DIR})
 
-IF (${CMAKE_GENERATOR} STREQUAL "Xcode")
+# Workaround for Xcode generator escaping issue, see
+# - https://cmake.org/cmake/help/v3.5/release/3.5.html#deprecated-and-removed-features
+IF (CMAKE_VERSION VERSION_LESS 3.5 AND ${CMAKE_GENERATOR} STREQUAL "Xcode")
     ADD_DEFINITIONS("-D TEST_DIR=\"\\\\\"${TESTDATA_SOURCE_DIR}\\\\\"\"")
-ELSE (${CMAKE_GENERATOR} STREQUAL "Xcode")
+ELSE (CMAKE_VERSION VERSION_LESS 3.5 AND ${CMAKE_GENERATOR} STREQUAL "Xcode")
     ADD_DEFINITIONS("-D TEST_DIR=\"\\\"${TESTDATA_SOURCE_DIR}\\\"\"")
-ENDIF (${CMAKE_GENERATOR} STREQUAL "Xcode")
+ENDIF (CMAKE_VERSION VERSION_LESS 3.5 AND ${CMAKE_GENERATOR} STREQUAL "Xcode")
 
 IF(NOT ${USE_RELATIVE_TEST_DIR})
     # Check if data exists

